// tera-data-parser/proto/tera.js
const Proto = require('protojs'),
	{ Customize, SkillID, Vec3 } = require('../types')

const MULT_INT16_TO_RAD = 1 / 0x8000 * Math.PI,
	MULT_RAD_TO_INT16 = 1 / Math.PI * 0x8000

const SKILL_ID_32 = {
	read(data) {
		const raw = data.dv.getUint32(data.pos, true),
			type = raw >> 26 & 0xf,
			npc = Boolean(raw & 0x40000000)

		data.pos += 4

		return new SkillID({
			id: raw & (npc ? 0xffff : 0x3ffffff),
			huntingZoneId: npc ? (raw >> 16) & 0x3ff : 0,
			type,
			npc,
			reserved: raw >> 31
		})
	},
	write(data, val) {
		if(val == null) data.buf.fill(0, data.pos, data.pos += 8)
		else {
			if(typeof val === 'number') val = { type: 1, id: val }

			let raw = (Number(val.id) || 0) & (val.npc ? 0xffff : 0x3ffffff)
			if(val.npc) raw |= (val.huntingZoneId & 0x3ff) << 16
			raw |= (val.type & 0xf) << 26
			raw |= (val.npc & 1) << 30
			raw |= (val.reserved & 1) << 31

			data.dv.setUint32(data.pos, raw, true)
			data.pos += 4
		}
	},
	length: 4
}

const SKILL_ID_64 = {
	read(data) {
		const raw = data.dv.getBigUint64(data.pos, true),
			type = Number(raw >> 28n & 0xfn),
			npc = Boolean(raw & 0x0100000000n)

		data.pos += 8

		return new SkillID({
			id: Number(raw & (npc ? 0xffffn : 0xfffffffn)),
			huntingZoneId: npc ? Number(raw >> 16n & 0xfffn) : 0,
			type,
			npc,
			reserved: Number(raw >> 33n)
		})
	},
	write(data, val) {
		if(val == null) data.buf.fill(0, data.pos, data.pos += 8)
		else {
			if(typeof val === 'number') val = { type: 1, id: val }

			let raw = BigInt((Number(val.id) || 0) & (val.npc ? 0xffff : 0xfffffff))
			if(val.npc) raw |= BigInt(val.huntingZoneId & 0xfff) << 16n
			raw |= BigInt(val.type & 0xf) << 28n
			raw |= BigInt(val.npc & 1) << 32n
			raw |= BigInt(val.reserved & 1) << 33n

			data.dv.setBigUint64(data.pos, raw, true)
			data.pos += 8
		}
	},
	length: 8
}

class ProtoTera extends Proto {
	constructor(gameVersion) {
		super(gameVersion, true)

		Object.assign(this, {
			bool: {
				compileRead(compile, id) { compile.str += `${id} = !!data.dv.getUint8(data.pos++)\n` },
				compileWrite(compile, id) { compile.str += `data.dv.setUint8(data.pos++, ${id} ? 1 : 0)\n` },
				length: 1
			},

			byte: Proto.StandardType('Uint8'),
			int16: Proto.StandardType('Int16'),
			uint16: Proto.StandardType('Uint16'),
			int32: Proto.StandardType('Int32'),
			uint32: Proto.StandardType('Uint32'),
			int64: Proto.StandardType('BigInt64'),
			uint64: Proto.StandardType('BigUint64'),
			float: Proto.StandardType('Float32'),
			double: Proto.StandardType('Float64'),

			vec3: {
				read(data) {
					const val = new Vec3(
						data.dv.getFloat32(data.pos, true),
						data.dv.getFloat32(data.pos + 4, true),
						data.dv.getFloat32(data.pos + 8, true)
					)
					data.pos += 12
					return val
				},
				compileWrite(compile, id) {
					compile.str += `if(${id} == null) data.buf.fill(0, data.pos, data.pos += 12)\n`
					compile.str += `else {\n`
					compile.write(['float'], `${id}.x`)
					compile.write(['float'], `${id}.y`)
					compile.write(['float'], `${id}.z`)
					compile.str += `}\n`
				},
				length: 12
			},

			vec3fa: {
				read(data) {
					const val = new Vec3(
						data.dv.getFloat32(data.pos, true) * MULT_INT16_TO_RAD,
						data.dv.getFloat32(data.pos + 4, true) * MULT_INT16_TO_RAD,
						data.dv.getFloat32(data.pos + 8, true) * MULT_INT16_TO_RAD
					)
					data.pos += 12
					return val
				},
				compileWrite(compile, id) {
					compile.str += `if(${id} == null) data.buf.fill(0, data.pos, data.pos += 12)\n`
					compile.str += `else {\n`
					compile.write(['float'], `Math.round(${id}.x * ${MULT_RAD_TO_INT16} % 0x10000)`)
					compile.write(['float'], `Math.round(${id}.y * ${MULT_RAD_TO_INT16} % 0x10000)`)
					compile.write(['float'], `Math.round(${id}.z * ${MULT_RAD_TO_INT16} % 0x10000)`)
					compile.str += `}\n`
				},
				length: 12
			},

			angle: {
				compileRead(compile, id) {
					compile.str += `${id} = data.dv.getInt16(data.pos, true) * ${MULT_INT16_TO_RAD}\n`
					compile.str += `data.pos += 2\n`
				},
				compileWrite(compile, id) { compile.write(['int16'], `Math.round(${id} * ${MULT_RAD_TO_INT16})`) },
				length: 2
			},

			skillid: gameVersion < 74 ? SKILL_ID_32 : SKILL_ID_64,

			customize: {
				read(data) {
					const val = new Customize(data.dv.getBigUint64(data.pos, true))
					data.pos += 8
					return val
				},
				write(data, val) {
					if(val == null) data.buf.fill(0, data.pos, data.pos += 8)
					else {
						if(typeof val === 'bigint') val = new Customize(val)
						data.dv.setUint8(data.pos, val.unk)
						data.dv.setUint8(data.pos + 1, val.skinColor)
						data.dv.setUint8(data.pos + 2, val.faceStyle)
						data.dv.setUint8(data.pos + 3, val.faceDecal)
						data.dv.setUint8(data.pos + 4, val.hairStyle)
						data.dv.setUint8(data.pos + 5, val.hairColor)
						data.dv.setUint8(data.pos + 6, val.voice)
						data.dv.setUint8(data.pos + 7, val.tattoos)
						data.pos += 8
					}
				},
				length: 8
			},

			array: {
				compileRead(compile, id, subType) {
					const size = compile.local(),
						pointer = compile.local(),
						rtnPos = compile.local(),
						temp = compile.local()

					compile.read(['uint16'], `let ${size}`)
					compile.read(['uint16'], `let ${pointer}`)
					compile.str += `const ${rtnPos} = data.pos\n`
					compile.str += `${id} = []\n`

					compile.str += `for(let ${temp}; ${size}--;) {\n`
					// Sanity check
					compile.str += `if(data.pos > ${pointer}) throw Error(\`Array element pointer is backwards ($\{data.pos - 4} > $\{${pointer}})\`)\n`
					// Jump to next element
					compile.str += `data.pos = ${pointer}\n`
					// Pointer to current element (for error check)
					compile.str += `if(data.dv.getUint16(data.pos, true) !== data.pos) throw Error(\`Invalid array element @ $\{data.pos}\`)\n`
					compile.str += `data.pos += 2\n`
					// Pointer to next element
					compile.read(['uint16'], pointer)
					// Element data
					if(subType) compile.read(subType, temp)
					compile.str += `${id}.push(${temp})\n`
					compile.str += `}\n`

					// Sanity check
					compile.str += `if(${pointer}) throw Error('Unterminated array')\n`
					compile.str += `data.pos = ${rtnPos}\n`
				},

				compileWrite(compile, id, subType) {
					compile.str += `if(!Array.isArray(${id}) && ${id} != null) throw TypeError('${id} must be an Array, null, or undefined')\n`
					id = compile.makeLocal(id, '[]')

					const ptrPos = compile.local(),
						i = compile.local()

					compile.write(['uint16'], `${id}.length`)
					compile.str += `let ${ptrPos} = data.pos\n`
					compile.write(['uint16'], '0')

					const main = compile.popState()
					compile.str += `for(let ${i} = 0; ${i} < ${id}.length; ${i}++) {\n`
					// Write current position to previous pointer
					compile.str += `data.dv.setUint16(${ptrPos}, data.pos, true)\n`
					// Pointer to current element (for error check)
					compile.write(['uint16'], 'data.pos')
					// Pointer to next element
					compile.str += `${ptrPos} = data.pos\n`
					compile.write(['uint16'], '0')
					// Element data
					if(subType) compile.write(subType, `${id}[${i}]`)
					compile.flushState()
					compile.str += `}\n`
					compile.pushState(main)
				},

				compileLength(compile, id, subType) {
					const i = compile.local(),
						length = subType ? compile.length(subType, `${id}[${i}]`) : 0

					compile.str += `if(!Array.isArray(${id}) && ${id} != null) throw TypeError('${id} must be an Array, null, or undefined')\n`
					id = compile.makeLocal(id, '[]')

					if(typeof length === 'number')
						compile.str += `len += 4 + ${4 + length}*${id}.length\n`
					else {
						compile.str += `len += 4 + 4*${id}.length\n`
						compile.str += `for(let ${i} = 0; ${i} < ${id}.length; ${i}++) {\n`
						compile.str += length
						compile.str += `}\n`
					}

					const str = compile.str
					compile.str = ''
					return str
				}
			},

			string: {
				compileRead(compile, id) {
					const pointer = compile.local(),
						rtnPos = compile.local()

					compile.read(['uint16'], `let ${pointer}`)

					// Sanity check
					compile.str += `if(data.pos > ${pointer}) throw Error(\`String pointer is backwards ($\{data.pos - 2} > $\{${pointer}})\`)\n`

					compile.str += `const ${rtnPos} = data.pos\n`
					compile.str += `data.pos = ${pointer}\n`
					// Find string terminator
					compile.str += `while(data.dv.getUint16(${pointer}, true)) ${pointer} += 2\n`
					compile.str += `${id} = ${pointer} === data.pos ? '' : data.buf.toString('ucs2', data.pos, ${pointer})\n`
					compile.str += `data.pos = ${rtnPos}\n`
				},

				compileWrite(compile, id) {
					compile.str += `if(typeof ${id} !== 'string' && ${id} != null) throw TypeError('${id} must be a string, null, or undefined')\n`
					id = compile.makeLocal(id, `''`)

					const ptrPos = compile.local()

					compile.str += `const ${ptrPos} = data.pos\n`
					compile.str += `data.pos += 2\n`

					const main = compile.popState()
					compile.str += `data.dv.setUint16(${ptrPos}, data.pos, true)\n`
					compile.str += `data.buf.fill(${id} + '\\0', data.pos, data.pos += 2 + 2*${id}.length, 'ucs2')\n`
					compile.pushState(main)
				},

				compileLength(compile, id) {
					compile.str += `if(typeof ${id} !== 'string' && ${id} != null) throw TypeError('${id} must be a string, null, or undefined')\n`
					id = compile.makeLocal(id, `''`)

					compile.str += `len += 4 + 2*(${id} || '').length\n`

					const str = compile.str
					compile.str = ''
					return str
				}
			},

			bytes: {
				compileRead(compile, id) {
					const pointer = compile.local(),
						size = compile.local(),
						rtnPos = compile.local()

					compile.read(['uint16'], `const ${pointer}`)
					compile.read(['uint16'], `const ${size}`)
					compile.str += `const ${rtnPos} = data.pos\n`

					// Sanity check
					compile.str += `if(${size} && data.pos > ${pointer}) throw Error(\`Bytes pointer is backwards ($\{data.pos - 4} > $\{${pointer}})\`)\n`

					compile.str += `data.pos = ${pointer}\n`
					compile.str += `${id} = Buffer.from(data.buf.slice(data.pos, data.pos + ${size}))\n`
					compile.str += `data.pos = ${rtnPos}\n`
				},

				compileWrite(compile, id) {
					compile.str += `if(!(${id} instanceof Buffer) && ${id} != null) throw TypeError('${id} must be a Buffer, null, or undefined')\n`
					id = compile.makeLocal(id, `''`)

					const ptrPos = compile.local()

					compile.str += `const ${ptrPos} = data.pos\n`
					compile.str += `data.pos += 2\n`
					compile.write(['uint16'], `${id}.length`)

					const main = compile.popState()
					compile.str += `if(!${id}.length) data.dv.setUint16(${ptrPos}, data.pos, true)\n`
					compile.str += `else {\n`
					compile.str += `data.dv.setUint16(${ptrPos}, data.pos, true)\n`
					compile.str += `${id}.copy(data.buf, data.pos)\n`
					compile.str += `data.pos += ${id}.length\n`
					compile.str += `}\n`
					compile.pushState(main)
				},

				compileLength(compile, id) {
					compile.str += `if(!(${id} instanceof Buffer) && ${id} != null) throw TypeError('${id} must be a Buffer, null, or undefined')\n`
					id = compile.makeLocal(id, `''`)

					compile.str += `len += 4 + (${id} || '').length\n`

					const str = compile.str
					compile.str = ''
					return str
				}
			}
		})
	}
}

module.exports = ProtoTera
